src/main/java/org/openhab/tools/analysis/checkstyle/readme/MarkDownVisitor.java

```diff
@@ -0,0 +1,336 @@
+     */
+    private void validateCodeSection(String codeBlock) {
+        int codeStartingLineNumber = 0;
+        int codeEndingLineNumber = 0;
+
+        // Splitting the String in case it is multiline, else the array would
+        // have only one element
+        String codeSectionLines[] = codeBlock.split("\\r?\\n");
+        if (!codeBlock.trim().isEmpty()) {
+            codeStartingLineNumber = callBack.findLineNumber(lines, codeSectionLines[0], currentLinePointer);
+            // if findLineNumber returns -1
+            if (codeStartingLineNumber >= 0) {
+                codeStartingLineNumber--;
+                currentLinePointer = codeStartingLineNumber;
+                // Start from the line above the code section
+                verifyBeforeCodeSection(codeStartingLineNumber);
+                codeEndingLineNumber = codeStartingLineNumber + codeSectionLines.length;
+                codeEndingLineNumber++;
+                currentLinePointer = codeEndingLineNumber;
+                verifyAfterCodeSection(codeEndingLineNumber);
+            } else {
+                logger.error("A code section wasn't processed properly! {}", codeBlock);
+            }
+        } else {
+            callBack.log(currentLinePointer, EMPTY_CODE_BLOCK_WARNING);
+        }
+    }
+
+    private void verifyBeforeCodeSection(int codeStartingLineNumber) {
+        if (codeStartingLineNumber == 1) {
+            callBack.log(codeStartingLineNumber, EMPTY_LINE_BEFORE_CODE_MSG);
+        } else {
+            // -2 because the line before the code is occupied with code section opening- ```
+            int lineBeforeCodeSection = codeStartingLineNumber - 2;
+            boolean isLineBeforeCodeSectionEmpty = lines.get(lineBeforeCodeSection).trim().isEmpty();
+            if (!isLineBeforeCodeSectionEmpty) {
+                callBack.log(codeStartingLineNumber, EMPTY_LINE_BEFORE_CODE_MSG);
+            }
+        }
+    }
+
+    private void verifyAfterCodeSection(int codeEndingLineNumber) {
+        boolean isCodeSectionAtEndOfFile = (codeEndingLineNumber == lines.size());
+        // There is another check that logs errors if there is not an empty line at the end of file
+        // named NewLineAtEndOfFileCheck
+        if (!isCodeSectionAtEndOfFile) {
+            if (!lines.get(codeEndingLineNumber).trim().isEmpty()) {
+                callBack.log(codeEndingLineNumber, EMPTY_LINE_AFTER_CODE_MSG);
+            }
+        }
+    }
+
+    /**
+     * Common method for processing visited {@link BulletList } and {@link OrderedList}
+     * 
+     * @param listBlock - a block type which is common parent of {@link BulletList } and {@link OrderedList}
+     */
+    private void processListBlock(ListBlock listBlock) {
+        String firstLineOfList = getFirstLineInList(listBlock);
+        String lastLineOfList = getLastLineInList(listBlock);
+        if (lastLineOfList == null) {
```

Comment:
Seems logical but since these two methods are independent from one another I don't have access to firstLineOfList and it seemed like a good decision to check it in this method.